// @fluentui/react-northstar/dist/es/themes/teams/components/Card/cardVariables.js
import { pxToRem } from '../../../../utils';
export default (function (siteVars) {
  return {
    borderStyle: 'solid',
    borderColor: siteVars.colors.black,
    borderWidth: siteVars.borderWidth,
    borderRadius: pxToRem(4),
    boxShadow: undefined,
    boxShadowHover: siteVars.shadowLevel3,
    padding: pxToRem(10),
    compactPadding: pxToRem(0),
    previewMargin: "0 0 ".concat(pxToRem(10), " 0"),
    headerMargin: "0 0 ".concat(pxToRem(10), " 0"),
    bodyMargin: "0 0 ".concat(pxToRem(10), " 0"),
    footerMargin: "0 0 ".concat(pxToRem(10), " 0"),
    fittedPreviewMargin: pxToRem(0),
    fittedHeaderMargin: pxToRem(0),
    fittedBodyMargin: pxToRem(0),
    fittedFooterMargin: pxToRem(0),
    previewMarginHorizontal: "0 ".concat(pxToRem(10), " 0 0"),
    topControlsTop: pxToRem(10),
    topControlsRight: pxToRem(0),
    // TODO: update with latest values from design
    width: pxToRem(300),
    height: '100%',
    sizeSmallWidth: pxToRem(200),
    sizeSmallHeight: '100%',
    sizeSmallPadding: pxToRem(5),
    sizeLargeWidth: pxToRem(500),
    sizeLargeHeight: '100%',
    sizeLargePadding: pxToRem(20),
    fluidWidth: '100%',
    fluidHeight: '100%'
  };
});
