// @fluentui/react-northstar/dist/dts/src/components/Box/Box.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_bindings_1 = require("@fluentui/react-bindings");
var React = require("react");
// @ts-ignore
var react_fela_1 = require("react-fela");
var utils_1 = require("../../utils");
var types_1 = require("../../types");
var Box = function (props) {
    var context = React.useContext(react_fela_1.ThemeContext);
    var _a = react_bindings_1.useTelemetry(Box.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, className = props.className, design = props.design, styles = props.styles, variables = props.variables, children = props.children, content = props.content;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: Box.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(Box.displayName, {
        className: Box.className,
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var unhandledProps = react_bindings_1.useUnhandledProps(Box.handledProps, props);
    var ElementType = react_bindings_1.getElementType(props);
    var result = (React.createElement(ElementType, tslib_1.__assign({}, getA11Props('root', tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, utils_1.rtlTextContainer.getAttributes({ forElements: [children, content] })), { className: classes.root }), unhandledProps))), utils_1.childrenExist(children) ? children : content));
    setEnd();
    return result;
};
Box.className = 'ui-box';
Box.displayName = 'Box';
Box.propTypes = utils_1.commonPropTypes.createCommon();
Box.handledProps = Object.keys(Box.propTypes);
Box.create = utils_1.createShorthandFactory({ Component: Box });
/**
 * A Box is a basic component, commonly used for slots in other Fluent UI components.
 * By default it just renders a `div`.
 */
exports.default = types_1.withSafeTypeForAs(Box);
