// @fluentui/react-bindings/dist/commonjs/telemetry/useTelemetry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getTelemetry = void 0;

var _useComposeOptions = _interopRequireDefault(require("../compose/useComposeOptions"));

var getTelemetry = function getTelemetry(displayName, telemetry) {
  var start = -1;
  var end = -1;

  var setStart = function setStart() {
    start = telemetry && telemetry.enabled ? performance.now() : -1;
  };

  var setEnd = function setEnd() {
    if (telemetry && telemetry.enabled && start !== -1) {
      end = performance.now();
      var duration = end - start;

      if (telemetry.performance[displayName]) {
        telemetry.performance[displayName].count++;
        telemetry.performance[displayName].msTotal += duration;
        telemetry.performance[displayName].msMin = Math.min(duration, telemetry.performance[displayName].msMin);
        telemetry.performance[displayName].msMax = Math.max(duration, telemetry.performance[displayName].msMax);
      } else {
        telemetry.performance[displayName] = {
          count: 1,
          msTotal: duration,
          msMin: duration,
          msMax: duration
        };
      }
    }
  };

  return {
    setStart: setStart,
    setEnd: setEnd
  };
};

exports.getTelemetry = getTelemetry;

var useTelemetry = function useTelemetry(displayName, telemetry) {
  var composeOptions = (0, _useComposeOptions.default)();
  return getTelemetry((composeOptions === null || composeOptions === void 0 ? void 0 : composeOptions.displayNames[0]) || displayName, telemetry);
};

var _default = useTelemetry;
exports.default = _default;
