// @fluentui/react-northstar/dist/commonjs/components/List/ListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@fluentui/accessibility");

var _reactBindings = require("@fluentui/react-bindings");

var _reactContextSelector = require("@fluentui/react-context-selector");

var _classnames = _interopRequireDefault(require("classnames"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _reactFela = require("react-fela");

var _Box = _interopRequireDefault(require("../Box/Box"));

var _types = require("../../types");

var _utils = require("../../utils");

var _listContext = require("./listContext");

// @ts-ignore
var ListItem = function ListItem(props) {
  var context = React.useContext(_reactFela.ThemeContext);

  var _useTelemetry = (0, _reactBindings.useTelemetry)(ListItem.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var accessibility = props.accessibility,
      className = props.className,
      content = props.content,
      contentMedia = props.contentMedia,
      design = props.design,
      endMedia = props.endMedia,
      header = props.header,
      important = props.important,
      headerMedia = props.headerMedia,
      media = props.media,
      styles = props.styles;
  var parentProps = (0, _reactContextSelector.useContextSelectors)(_listContext.ListContext, {
    debug: function debug(v) {
      return v.debug;
    },
    navigable: function navigable(v) {
      return v.navigable;
    },
    selectable: function selectable(v) {
      return v.selectable;
    },
    truncateContent: function truncateContent(v) {
      return v.truncateContent;
    },
    truncateHeader: function truncateHeader(v) {
      return v.truncateHeader;
    },
    variables: function variables(v) {
      return v.variables;
    },
    onItemClick: function onItemClick(v) {
      return v.onItemClick;
    },
    selected: function selected(v) {
      return v.selectedIndex === props.index;
    }
  });
  var _props$debug = props.debug,
      debug = _props$debug === void 0 ? parentProps.debug : _props$debug,
      _props$navigable = props.navigable,
      navigable = _props$navigable === void 0 ? parentProps.navigable : _props$navigable,
      _props$selectable = props.selectable,
      selectable = _props$selectable === void 0 ? parentProps.selectable : _props$selectable,
      _props$selected = props.selected,
      selected = _props$selected === void 0 ? parentProps.selected : _props$selected,
      _props$truncateConten = props.truncateContent,
      truncateContent = _props$truncateConten === void 0 ? parentProps.truncateContent : _props$truncateConten,
      _props$truncateHeader = props.truncateHeader,
      truncateHeader = _props$truncateHeader === void 0 ? parentProps.truncateHeader : _props$truncateHeader,
      _props$variables = props.variables,
      variables = _props$variables === void 0 ? parentProps.variables : _props$variables;
  var getA11Props = (0, _reactBindings.useAccessibility)(accessibility, {
    debugName: ListItem.displayName,
    actionHandlers: {
      performClick: function performClick(e) {
        e.preventDefault();
        handleClick(e);
      }
    },
    mapPropsToBehavior: function mapPropsToBehavior() {
      return {
        navigable: navigable,
        selectable: selectable,
        selected: selected
      };
    },
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(ListItem.displayName, {
    className: ListItem.className,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        debug: debug,
        navigable: navigable,
        important: important,
        selectable: selectable,
        selected: selected,
        truncateContent: truncateContent,
        truncateHeader: truncateHeader,
        hasContent: !!content,
        hasContentMedia: !!contentMedia,
        hasHeader: !!header,
        hasHeaderMedia: !!headerMedia
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes,
      resolvedStyles = _useStyles.styles;

  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(ListItem.handledProps, props);

  var handleClick = function handleClick(e) {
    (0, _invoke2.default)(props, 'onClick', e, props);
    parentProps.onItemClick(e, props.index);
  };

  var contentElement = _Box.default.create(content, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.content,
        styles: resolvedStyles.content
      };
    }
  });

  var contentMediaElement = _Box.default.create(contentMedia, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.contentMedia,
        styles: resolvedStyles.contentMedia
      };
    }
  });

  var headerElement = _Box.default.create(header, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.header,
        styles: resolvedStyles.header
      };
    }
  });

  var headerMediaElement = _Box.default.create(headerMedia, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.headerMedia,
        styles: resolvedStyles.headerMedia
      };
    }
  });

  var endMediaElement = _Box.default.create(endMedia, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.endMedia,
        styles: resolvedStyles.endMedia
      };
    }
  });

  var mediaElement = _Box.default.create(media, {
    defaultProps: function defaultProps() {
      return {
        className: ListItem.slotClassNames.media,
        styles: resolvedStyles.media
      };
    }
  });

  var element = React.createElement(ElementType, getA11Props('root', Object.assign({
    className: classes.root,
    onClick: handleClick
  }, unhandledProps)), mediaElement, React.createElement("div", {
    className: (0, _classnames.default)(ListItem.slotClassNames.main, classes.main)
  }, (headerElement || headerMediaElement) && React.createElement("div", {
    className: (0, _classnames.default)(ListItem.slotClassNames.headerWrapper, classes.headerWrapper)
  }, headerElement, headerMediaElement), (contentElement || contentMediaElement) && React.createElement("div", {
    className: (0, _classnames.default)(ListItem.slotClassNames.contentWrapper, classes.contentWrapper)
  }, contentElement, contentMediaElement)), endMediaElement);
  setEnd();
  return element;
};

ListItem.className = 'ui-list__item';
ListItem.displayName = 'ListItem';
ListItem.defaultProps = {
  as: 'li',
  accessibility: _accessibility.listItemBehavior
};
ListItem.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon({
  content: false
}), {
  contentMedia: PropTypes.any,
  content: PropTypes.any,
  debug: PropTypes.bool,
  header: PropTypes.any,
  endMedia: PropTypes.any,
  headerMedia: PropTypes.any,
  important: PropTypes.bool,
  media: PropTypes.any,
  selectable: PropTypes.bool,
  navigable: PropTypes.bool,
  index: PropTypes.number,
  selected: PropTypes.bool,
  truncateContent: PropTypes.bool,
  truncateHeader: PropTypes.bool,
  onClick: PropTypes.func
});
ListItem.handledProps = Object.keys(ListItem.propTypes);
ListItem.slotClassNames = {
  header: "".concat(ListItem.className, "__header"),
  headerMedia: "".concat(ListItem.className, "__headerMedia"),
  headerWrapper: "".concat(ListItem.className, "__headerWrapper"),
  main: "".concat(ListItem.className, "__main"),
  content: "".concat(ListItem.className, "__content"),
  contentMedia: "".concat(ListItem.className, "__contentMedia"),
  contentWrapper: "".concat(ListItem.className, "__contentWrapper"),
  media: "".concat(ListItem.className, "__media"),
  endMedia: "".concat(ListItem.className, "__endMedia")
};
ListItem.create = (0, _utils.createShorthandFactory)({
  Component: ListItem,
  mappedProp: 'content'
});
/**
 * A ListItem contains a single piece of content within a List.
 */

var _default = (0, _types.withSafeTypeForAs)(ListItem);

exports.default = _default;
