// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Divider/dividerStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _get2 = _interopRequireDefault(require("lodash/get"));

var _utils = require("../../../../utils");

var beforeAndAfter = function beforeAndAfter(size, variables, colors, props) {
  return Object.assign({
    content: '""',
    flex: 1
  }, props.vertical ? {
    width: "".concat(size + 1, "px"),
    height: '100%'
  } : {
    height: "".concat(size + 1, "px")
  }, {
    background: (0, _get2.default)(colors, 'foreground', variables.dividerColor)
  });
};

var dividerStyles = {
  root: function root(_ref) {
    var props = _ref.props,
        variables = _ref.variables;
    var children = props.children,
        color = props.color,
        fitted = props.fitted,
        size = props.size,
        important = props.important,
        content = props.content,
        vertical = props.vertical;
    var colors = variables.colorScheme[color];
    return Object.assign({
      color: (0, _get2.default)(colors, 'foreground', variables.textColor),
      display: 'flex',
      alignItems: 'center'
    }, !fitted && {
      padding: vertical ? "0 ".concat(variables.dividerPadding) : "".concat(variables.dividerPadding, " 0")
    }, {}, important && {
      fontWeight: variables.importantFontWeight
    }, {}, vertical && {
      height: '100%'
    }, {}, (0, _utils.childrenExist)(children) || content ? {
      textAlign: 'center',
      fontSize: (0, _utils.pxToRem)(12 + size),
      lineHeight: variables.textLineHeight,
      '::before': Object.assign({}, beforeAndAfter(size, variables, colors, props), {
        marginRight: (0, _utils.pxToRem)(20)
      }),
      '::after': Object.assign({}, beforeAndAfter(size, variables, colors, props), {
        marginLeft: (0, _utils.pxToRem)(20)
      })
    } : {
      '::before': Object.assign({}, beforeAndAfter(size, variables, colors, props))
    });
  }
};
var _default = dividerStyles;
exports.default = _default;
