// @fluentui/react-northstar/dist/es/components/SplitButton/SplitButton.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _invoke from "lodash/invoke";
import { splitButtonBehavior } from '@fluentui/accessibility';
import * as customPropTypes from '@fluentui/react-proptypes';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { ALIGNMENTS, POSITIONS } from '../../utils/positioner';
import { withSafeTypeForAs } from '../../types';
import { isFromKeyboard, commonPropTypes, AutoControlledComponent } from '../../utils';
import SplitButtonToggle from './SplitButtonToggle';
import Button from '../Button/Button';
import MenuButton from '../MenuButton/MenuButton';
import { Ref } from '@fluentui/react-component-ref';

var SplitButton =
/*#__PURE__*/
function (_AutoControlledCompon) {
  _inherits(SplitButton, _AutoControlledCompon);

  function SplitButton() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, SplitButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(SplitButton)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "targetRef", React.createRef());

    _defineProperty(_assertThisInitialized(_this), "handleMenuButtonOverrides", function (predefinedProps) {
      return {
        onMenuItemClick: function onMenuItemClick(e, menuItemProps) {
          _this.setState({
            open: false
          });

          _invoke(_this.props, 'onOpenChange', e, Object.assign({}, _this.props, {
            open: false
          }));

          _invoke(predefinedProps, 'onMenuItemClick', e, menuItemProps);

          _invoke(_this.props, 'onMenuItemClick', e, menuItemProps);
        },
        onOpenChange: function onOpenChange(e, popupProps) {
          e.stopPropagation();

          _this.setState({
            open: popupProps.open
          });

          _invoke(_this.props, 'onOpenChange', e, Object.assign({}, _this.props, {
            open: popupProps.open
          }));
        }
      };
    });

    _defineProperty(_assertThisInitialized(_this), "handleMenuButtonTriggerOverrides", function (predefinedProps) {
      return {
        onClick: function onClick(e, buttonProps) {
          _invoke(predefinedProps, 'onClick', e, buttonProps);

          _invoke(_this.props, 'onMainButtonClick', e, buttonProps);
        },
        onFocus: function onFocus(e, buttonProps) {
          _invoke(predefinedProps, 'onFocus', e, buttonProps);

          _this.setState({
            isFromKeyboard: isFromKeyboard()
          });
        }
      };
    });

    return _this;
  }

  _createClass(SplitButton, [{
    key: "getInitialAutoControlledState",
    value: function getInitialAutoControlledState() {
      return {
        isFromKeyboard: false,
        open: false
      };
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          button = _this$props.button,
          disabled = _this$props.disabled,
          menu = _this$props.menu,
          primary = _this$props.primary,
          secondary = _this$props.secondary,
          toggleButton = _this$props.toggleButton,
          size = _this$props.size,
          position = _this$props.position,
          align = _this$props.align,
          flipBoundary = _this$props.flipBoundary,
          overflowBoundary = _this$props.overflowBoundary,
          positionFixed = _this$props.positionFixed,
          offset = _this$props.offset,
          unstable_pinned = _this$props.unstable_pinned;
      var trigger = Button.create(button, {
        defaultProps: function defaultProps() {
          return {
            styles: styles.menuButton,
            primary: primary,
            secondary: secondary,
            disabled: disabled
          };
        },
        overrideProps: this.handleMenuButtonTriggerOverrides
      });
      return React.createElement(Ref, {
        innerRef: this.targetRef
      }, React.createElement(ElementType, _extends({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps), MenuButton.create({}, {
        defaultProps: function defaultProps() {
          return {
            accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.menuButton : undefined,
            menu: menu,
            // Opening is handled manually.
            on: [],
            open: _this2.state.open,
            trigger: trigger,
            target: _this2.targetRef.current,
            position: position,
            align: align,
            flipBoundary: flipBoundary,
            overflowBoundary: overflowBoundary,
            positionFixed: positionFixed,
            offset: offset,
            unstable_pinned: unstable_pinned
          };
        },
        overrideProps: this.handleMenuButtonOverrides
      }), SplitButtonToggle.create(toggleButton, {
        defaultProps: function defaultProps() {
          return Object.assign({
            disabled: disabled,
            primary: primary,
            secondary: secondary,
            size: size
          }, accessibility.attributes.toggleButton);
        },
        overrideProps: function overrideProps(predefinedProps) {
          return {
            onClick: function onClick(e, buttonProps) {
              _invoke(predefinedProps, 'onClick', e, buttonProps);

              _this2.setState(function (state) {
                var open = !state.open;

                _invoke(_this2.props, 'onOpenChange', e, Object.assign({}, _this2.props, {
                  open: open
                }));

                return {
                  open: open
                };
              });
            }
          };
        }
      })));
    }
  }]);

  return SplitButton;
}(AutoControlledComponent);
/**
 * A SplitButton enables users to take one of several related actions, one being dominant and rest being displayed in a menu.
 */


_defineProperty(SplitButton, "create", void 0);

_defineProperty(SplitButton, "displayName", 'SplitButton');

_defineProperty(SplitButton, "Toggle", SplitButtonToggle);

_defineProperty(SplitButton, "className", 'ui-splitbutton');

_defineProperty(SplitButton, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  content: false
}), {
  button: customPropTypes.itemShorthand,
  defaultOpen: PropTypes.bool,
  menu: PropTypes.oneOfType([customPropTypes.itemShorthand, PropTypes.arrayOf(customPropTypes.itemShorthandWithoutJSX)]),
  onMainButtonClick: PropTypes.func,
  onMenuItemClick: PropTypes.func,
  onOpenChange: PropTypes.func,
  open: PropTypes.bool,
  size: customPropTypes.size,
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  toggleButton: customPropTypes.itemShorthand,
  position: PropTypes.oneOf(POSITIONS),
  align: PropTypes.oneOf(ALIGNMENTS),
  flipBoundary: PropTypes.oneOfType([PropTypes.object, PropTypes.oneOf(['scrollParent', 'window', 'viewport'])]),
  overflowBoundary: PropTypes.oneOfType([PropTypes.object, PropTypes.oneOf(['scrollParent', 'window', 'viewport'])]),
  positionFixed: PropTypes.bool,
  offset: PropTypes.string,
  unstable_pinned: PropTypes.bool
}));

_defineProperty(SplitButton, "defaultProps", {
  accessibility: splitButtonBehavior,
  toggleButton: {},
  position: 'below',
  align: 'start'
});

_defineProperty(SplitButton, "autoControlledProps", ['open']);

export default withSafeTypeForAs(SplitButton);
