// @fluentui/react-northstar/dist/es/components/Table/Table.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _invoke from "lodash/invoke";
import _map from "lodash/map";
import { tableBehavior } from '@fluentui/accessibility';
import * as customPropTypes from '@fluentui/react-proptypes';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { commonPropTypes, UIComponent, applyAccessibilityKeyHandlers, childrenExist } from '../../utils';
import { mergeComponentVariables } from '@fluentui/styles';
import TableRow from './TableRow';
import TableCell from './TableCell';
import { withSafeTypeForAs } from '../../types';

var handleVariablesOverrides = function handleVariablesOverrides(variables) {
  return function (predefinedProps) {
    return {
      variables: mergeComponentVariables(variables, predefinedProps.variables)
    };
  };
};

var Table =
/*#__PURE__*/
function (_UIComponent) {
  _inherits(Table, _UIComponent);

  function Table() {
    _classCallCheck(this, Table);

    return _possibleConstructorReturn(this, _getPrototypeOf(Table).apply(this, arguments));
  }

  _createClass(Table, [{
    key: "renderRows",
    value: function renderRows(accessibility, variables) {
      var _this$props = this.props,
          rows = _this$props.rows,
          compact = _this$props.compact;
      return _map(rows, function (row, index) {
        var props = {
          compact: compact,
          onClick: function onClick(e, props) {
            _invoke(row, 'onClick', e, props);
          }
        };
        var overrideProps = handleVariablesOverrides(variables);
        return TableRow.create(row, {
          defaultProps: function defaultProps() {
            return Object.assign({}, props, {
              accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.row : undefined
            });
          },
          overrideProps: overrideProps
        });
      });
    }
  }, {
    key: "renderHeader",
    value: function renderHeader(accessibility, variables) {
      var _this$props2 = this.props,
          header = _this$props2.header,
          compact = _this$props2.compact;

      if (!header) {
        return null;
      }

      var headerRowProps = {
        header: true,
        compact: compact,
        className: Table.slotClassNames.header
      };
      var overrideProps = handleVariablesOverrides(variables);
      return TableRow.create(header, {
        defaultProps: function defaultProps() {
          return Object.assign({}, headerRowProps, {
            accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.row : undefined
          });
        },
        overrideProps: overrideProps
      });
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          variables = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var children = this.props.children;
      var hasChildren = childrenExist(children);
      return React.createElement(ElementType, _extends({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps, applyAccessibilityKeyHandlers(accessibility.keyHandlers.root, unhandledProps)), hasChildren && children, !hasChildren && this.renderHeader(accessibility, variables), !hasChildren && this.renderRows(accessibility, variables));
    }
  }]);

  return Table;
}(UIComponent);
/**
 * A Table is used to display data in tabular layout
 * * @accessibility
 * Implements ARIA [Data Grid](https://www.w3.org/TR/wai-aria-practices/#dataGrid) design pattern for presenting tabular information.
 * When gridcell contains actionable element, use [gridCellWithFocusableElementBehavior](/components/table/accessibility#grid-cell-with-focusable-element-behavior-ts). [More information available in aria documentation.](https://www.w3.org/TR/wai-aria-practices/#gridNav_focus)
 * Use [gridCellMultipleFocusableBehavior](/components/table/accessibility#gridCellMultipleFocusableBehavior), when gridcell contains:
 * \- editable content
 * \- multiple actionable elements
 * \- component that utilizes arrow keys in its navigation, like menu button, dropdown, radio group, slider, etc.
 * [More information available in aria documentation.](https://www.w3.org/TR/wai-aria-practices/#gridNav_inside)
 * @accessibilityIssues
 * [NVDA narrate table title(aria-label) twice](https://github.com/nvaccess/nvda/issues/10548)
 * [Accessibility DOM > Table > gridcell > when gridcell is focused, then selected state is send to reader](https://bugs.chromium.org/p/chromium/issues/detail?id=1030378)
 * [JAWS narrate grid name twice, once as table and second time as grid](https://github.com/FreedomScientific/VFO-standards-support/issues/346)
 * [JAWS doesn't narrate grid column name, when focus is on actionable element in the cell] (https://github.com/FreedomScientific/VFO-standards-support/issues/348)
 * [aria-sort is not output at child elements](https://github.com/FreedomScientific/VFO-standards-support/issues/319)
 * [VoiceOver not announcing rows correctly for a grid with presentation elements inside](https://bugs.chromium.org/p/chromium/issues/detail?id=1054424)
 * VoiceOver doesn't narrate aria-rowcount value in table or grid
 */


_defineProperty(Table, "displayName", 'Table');

_defineProperty(Table, "className", 'ui-table');

_defineProperty(Table, "Cell", TableCell);

_defineProperty(Table, "Row", TableRow);

_defineProperty(Table, "slotClassNames", {
  header: "".concat(Table.className, "__header")
});

_defineProperty(Table, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  content: false
}), {
  content: customPropTypes.every([customPropTypes.disallow(['children']), PropTypes.oneOfType([PropTypes.arrayOf(customPropTypes.nodeContent), customPropTypes.nodeContent])]),
  header: customPropTypes.itemShorthand,
  rows: customPropTypes.collectionShorthand,
  compact: PropTypes.bool
}));

_defineProperty(Table, "defaultProps", {
  as: 'div',
  accessibility: tableBehavior
});

export default withSafeTypeForAs(Table);
