// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Card/cardVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
exports.default = (function (siteVars) {
    return {
        borderStyle: 'solid',
        borderColor: siteVars.colors.black,
        borderWidth: siteVars.borderWidth,
        borderRadius: utils_1.pxToRem(4),
        boxShadow: undefined,
        boxShadowHover: siteVars.shadowLevel3,
        padding: utils_1.pxToRem(10),
        compactPadding: utils_1.pxToRem(0),
        previewMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        headerMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        bodyMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        footerMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        fittedPreviewMargin: utils_1.pxToRem(0),
        fittedHeaderMargin: utils_1.pxToRem(0),
        fittedBodyMargin: utils_1.pxToRem(0),
        fittedFooterMargin: utils_1.pxToRem(0),
        previewMarginHorizontal: "0 " + utils_1.pxToRem(10) + " 0 0",
        topControlsTop: utils_1.pxToRem(10),
        topControlsRight: utils_1.pxToRem(0),
        // TODO: update with latest values from design
        width: utils_1.pxToRem(300),
        height: '100%',
        sizeSmallWidth: utils_1.pxToRem(200),
        sizeSmallHeight: '100%',
        sizeSmallPadding: utils_1.pxToRem(5),
        sizeLargeWidth: utils_1.pxToRem(500),
        sizeLargeHeight: '100%',
        sizeLargePadding: utils_1.pxToRem(20),
        fluidWidth: '100%',
        fluidHeight: '100%',
    };
});
