// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Card/cardVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../../../utils");

var _default = function _default(siteVars) {
  return {
    borderStyle: 'solid',
    borderColor: siteVars.colors.black,
    borderWidth: siteVars.borderWidth,
    borderRadius: (0, _utils.pxToRem)(4),
    boxShadow: undefined,
    boxShadowHover: siteVars.shadowLevel3,
    padding: (0, _utils.pxToRem)(10),
    compactPadding: (0, _utils.pxToRem)(0),
    previewMargin: "0 0 ".concat((0, _utils.pxToRem)(10), " 0"),
    headerMargin: "0 0 ".concat((0, _utils.pxToRem)(10), " 0"),
    bodyMargin: "0 0 ".concat((0, _utils.pxToRem)(10), " 0"),
    footerMargin: "0 0 ".concat((0, _utils.pxToRem)(10), " 0"),
    fittedPreviewMargin: (0, _utils.pxToRem)(0),
    fittedHeaderMargin: (0, _utils.pxToRem)(0),
    fittedBodyMargin: (0, _utils.pxToRem)(0),
    fittedFooterMargin: (0, _utils.pxToRem)(0),
    previewMarginHorizontal: "0 ".concat((0, _utils.pxToRem)(10), " 0 0"),
    topControlsTop: (0, _utils.pxToRem)(10),
    topControlsRight: (0, _utils.pxToRem)(0),
    // TODO: update with latest values from design
    width: (0, _utils.pxToRem)(300),
    height: '100%',
    sizeSmallWidth: (0, _utils.pxToRem)(200),
    sizeSmallHeight: '100%',
    sizeSmallPadding: (0, _utils.pxToRem)(5),
    sizeLargeWidth: (0, _utils.pxToRem)(500),
    sizeLargeHeight: '100%',
    sizeLargePadding: (0, _utils.pxToRem)(20),
    fluidWidth: '100%',
    fluidHeight: '100%'
  };
};

exports.default = _default;
