// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Card/cardStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _getBorderFocusStyles = _interopRequireDefault(require("../../getBorderFocusStyles"));

var cardStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        theme = _ref.theme;
    var siteVariables = theme.siteVariables;
    var borderFocusStyles = (0, _getBorderFocusStyles.default)({
      variables: siteVariables,
      borderWidth: v.borderWidth,
      borderRadius: v.borderRadius
    });
    return Object.assign({
      display: 'flex',
      flexDirection: 'column',
      position: 'relative',
      overflow: 'hidden',
      padding: v.padding,
      width: v.width,
      height: v.height,
      boxShadow: v.boxShadow
    }, p.size === 'small' && {
      width: v.sizeSmallWidth,
      height: v.sizeSmallHeight,
      padding: v.sizeSmallPadding
    }, {}, p.size === 'large' && {
      width: v.sizeLargeWidth,
      height: v.sizeLargeHeight,
      padding: v.sizeLargePadding
    }, {}, p.fluid && {
      width: v.fluidWidth,
      height: v.fluidHeight
    }, {}, p.horizontal && {
      flexDirection: 'row'
    }, {}, p.compact && {
      padding: v.compactPadding
    }, {}, p.centered && {
      alignItems: 'center'
    }, {
      ':hover': {
        boxShadow: v.boxShadowHover
      },
      borderWidth: v.borderWidth,
      borderStyle: v.borderStyle,
      borderColor: v.borderColor,
      borderRadius: v.borderRadius
    }, borderFocusStyles);
  }
};
var _default = cardStyles;
exports.default = _default;
