// @fluentui/react-northstar/dist/commonjs/components/Button/Button.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@fluentui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _utils = require("../../utils");

var _Box = _interopRequireDefault(require("../Box/Box"));

var _Loader = _interopRequireDefault(require("../Loader/Loader"));

var _types = require("../../types");

var _ButtonGroup = _interopRequireDefault(require("./ButtonGroup"));

var _ButtonContent = _interopRequireDefault(require("./ButtonContent"));

var _reactBindings = require("@fluentui/react-bindings");

var _reactFela = require("react-fela");

// @ts-ignore
var Button = function Button(props) {
  var context = React.useContext(_reactFela.ThemeContext);

  var _useTelemetry = (0, _reactBindings.useTelemetry)(Button.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var accessibility = props.accessibility,
      active = props.active,
      as = props.as,
      children = props.children,
      content = props.content,
      icon = props.icon,
      loader = props.loader,
      disabled = props.disabled,
      iconPosition = props.iconPosition,
      loading = props.loading,
      text = props.text,
      primary = props.primary,
      inverted = props.inverted,
      size = props.size,
      iconOnly = props.iconOnly,
      fluid = props.fluid,
      circular = props.circular,
      className = props.className,
      styles = props.styles,
      variables = props.variables,
      design = props.design;
  var hasChildren = (0, _utils.childrenExist)(children);
  var getA11Props = (0, _reactBindings.useAccessibility)(accessibility, {
    debugName: Button.displayName,
    mapPropsToBehavior: function mapPropsToBehavior() {
      return {
        as: as,
        active: active,
        disabled: disabled,
        loading: loading
      };
    },
    actionHandlers: {
      performClick: function performClick(event) {
        event.preventDefault();
        handleClick(event);
      }
    },
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(Button.displayName, {
    className: Button.className,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        text: text,
        primary: primary,
        disabled: disabled,
        circular: circular,
        size: size,
        loading: loading,
        inverted: inverted,
        iconOnly: iconOnly,
        iconPosition: iconPosition,
        fluid: fluid,
        hasContent: !!content
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes,
      resolvedStyles = _useStyles.styles;

  var unhandledProps = (0, _reactBindings.useUnhandledProps)(Button.handledProps, props);
  var ElementType = (0, _reactBindings.getElementType)(props);

  var renderIcon = function renderIcon() {
    return _Box.default.create(icon, {
      defaultProps: function defaultProps() {
        return getA11Props('icon', {
          styles: resolvedStyles.icon
        });
      }
    });
  };

  var renderLoader = function renderLoader() {
    return _Loader.default.create(loader || {}, {
      defaultProps: function defaultProps() {
        return getA11Props('loader', {
          role: undefined,
          styles: resolvedStyles.loader
        });
      }
    });
  };

  var handleClick = function handleClick(e) {
    if (disabled) {
      e.preventDefault();
      return;
    }

    (0, _invoke2.default)(props, 'onClick', e, props);
  };

  var handleFocus = function handleFocus(e) {
    (0, _invoke2.default)(props, 'onFocus', e, props);
  };

  var result = React.createElement(ElementType, (0, _extends2.default)({}, _utils.rtlTextContainer.getAttributes({
    forElements: [children]
  }), getA11Props('root', Object.assign({
    onClick: handleClick,
    disabled: disabled,
    className: classes.root,
    onFocus: handleFocus
  }, unhandledProps))), hasChildren ? children : React.createElement(React.Fragment, null, loading && renderLoader(), iconPosition !== 'after' && renderIcon(), _ButtonContent.default.create(content, {
    defaultProps: function defaultProps() {
      return getA11Props('content', {
        as: 'span',
        size: size,
        styles: resolvedStyles.content
      });
    }
  }), iconPosition === 'after' && renderIcon()));
  setEnd();
  return result;
};

Button.defaultProps = {
  as: 'button',
  accessibility: _accessibility.buttonBehavior,
  size: 'medium'
};
Button.displayName = 'Button';
Button.className = 'ui-button';
Button.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon({
  content: 'shorthand'
}), {
  circular: PropTypes.bool,
  disabled: PropTypes.bool,
  fluid: PropTypes.bool,
  icon: customPropTypes.shorthandAllowingChildren,
  iconOnly: PropTypes.bool,
  iconPosition: PropTypes.oneOf(['before', 'after']),
  inverted: PropTypes.bool,
  loader: customPropTypes.itemShorthandWithoutJSX,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  text: PropTypes.bool,
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  size: customPropTypes.size
});
Button.handledProps = Object.keys(Button.propTypes);
Button.Group = _ButtonGroup.default;
Button.Content = _ButtonContent.default;
Button.create = (0, _utils.createShorthandFactory)({
  Component: Button,
  mappedProp: 'content'
});
/**
 * A Button enables users to take an action, such as submitting a form, opening a dialog, etc.
 *
 * @accessibility
 * Implements [ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button) design pattern.
 */

var _default = (0, _types.withSafeTypeForAs)(Button);

exports.default = _default;
