// @fluentui/react-northstar/dist/commonjs/components/Carousel/CarouselItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _accessibility = require("@fluentui/accessibility");

var _utils = require("../../utils");

var _accessibilityStyles = require("../../utils/accessibility/Styles/accessibilityStyles");

var _types = require("../../types");

var CarouselItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(CarouselItem, _UIComponent);

  function CarouselItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, CarouselItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(CarouselItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      arrowKeysNavigationStopPropagation: function arrowKeysNavigationStopPropagation(e) {
        // let event propagate, when it was invoke on the element where arrow keys should rotate carousel
        if (e.currentTarget !== e.target) {
          e.stopPropagation();
        }
      }
    });
    return _this;
  }

  (0, _createClass2.default)(CarouselItem, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          styles = _ref.styles,
          accessibility = _ref.accessibility,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content,
          itemPositionText = _this$props.itemPositionText;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps, (0, _utils.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), (0, _utils.childrenExist)(children) ? children : content, React.createElement("div", {
        className: CarouselItem.slotClassNames.itemPositionText,
        style: _accessibilityStyles.screenReaderContainerStyles
      }, itemPositionText));
    }
  }]);
  return CarouselItem;
}(_utils.UIComponent);

(0, _defineProperty2.default)(CarouselItem, "create", void 0);
(0, _defineProperty2.default)(CarouselItem, "displayName", 'CarouselItem');
(0, _defineProperty2.default)(CarouselItem, "className", 'ui-carousel__item');
(0, _defineProperty2.default)(CarouselItem, "propTypes", Object.assign({}, _utils.commonPropTypes.createCommon(), {
  active: PropTypes.bool,
  navigation: PropTypes.bool,
  itemPositionText: PropTypes.string
}));
(0, _defineProperty2.default)(CarouselItem, "defaultProps", {
  accessibility: _accessibility.carouselItemBehavior
});
(0, _defineProperty2.default)(CarouselItem, "slotClassNames", {
  itemPositionText: "".concat(CarouselItem.className, "__itemPositionText")
});
CarouselItem.create = (0, _utils.createShorthandFactory)({
  Component: CarouselItem,
  mappedProp: 'content'
});
/**
 * A Carousel displays data organised as a gallery.
 *
 * @accessibility
 * Implements [ARIA Carousel](https://www.w3.org/WAI/tutorials/carousels/structure/) design pattern.
 */

var _default = (0, _types.withSafeTypeForAs)(CarouselItem);

exports.default = _default;
