// @fluentui/react-northstar/dist/commonjs/components/Carousel/CarouselNavigation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@fluentui/accessibility");

var _styles = require("@fluentui/styles");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var _utils = require("../../utils");

var _types = require("../../types");

var _CarouselNavigationItem = _interopRequireDefault(require("./CarouselNavigationItem"));

var CarouselNavigation =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(CarouselNavigation, _UIComponent);

  function CarouselNavigation() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, CarouselNavigation);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(CarouselNavigation)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleItemOverrides", function (variables) {
      return function (predefinedProps) {
        return {
          onClick: function onClick(e, itemProps) {
            (0, _invoke2.default)(_this.props, 'onItemClick', e, itemProps);
            (0, _invoke2.default)(predefinedProps, 'onClick', e, itemProps);
          },
          variables: (0, _styles.mergeComponentVariables)(variables, predefinedProps.variables)
        };
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderItems", function (variables, accessibility) {
      var _this$props = _this.props,
          activeIndex = _this$props.activeIndex,
          iconOnly = _this$props.iconOnly,
          items = _this$props.items,
          primary = _this$props.primary,
          secondary = _this$props.secondary,
          vertical = _this$props.vertical;
      return (0, _map2.default)(items, function (item, index) {
        return _CarouselNavigationItem.default.create(item, {
          defaultProps: function defaultProps() {
            return {
              active: index === activeIndex,
              iconOnly: iconOnly,
              index: index,
              primary: primary,
              secondary: secondary,
              vertical: vertical,
              accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.item : undefined
            };
          },
          overrideProps: _this.handleItemOverrides(variables)
        });
      });
    });
    return _this;
  }

  (0, _createClass2.default)(CarouselNavigation, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          variables = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var children = this.props.children;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, _utils.rtlTextContainer.getAttributes({
        forElements: [children]
      }), unhandledProps), (0, _utils.childrenExist)(children) ? children : this.renderItems(variables, accessibility));
    }
  }]);
  return CarouselNavigation;
}(_utils.UIComponent);

(0, _defineProperty2.default)(CarouselNavigation, "displayName", 'CarouselNavigation');
(0, _defineProperty2.default)(CarouselNavigation, "className", 'ui-carousel__navigation');
(0, _defineProperty2.default)(CarouselNavigation, "create", void 0);
(0, _defineProperty2.default)(CarouselNavigation, "propTypes", Object.assign({}, _utils.commonPropTypes.createCommon({
  content: false
}), {
  activeIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  iconOnly: PropTypes.bool,
  items: customPropTypes.collectionShorthand,
  onItemClick: PropTypes.func,
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  vertical: PropTypes.bool
}));
(0, _defineProperty2.default)(CarouselNavigation, "defaultProps", {
  accessibility: _accessibility.tabListBehavior,
  as: 'ul'
});
CarouselNavigation.create = (0, _utils.createShorthandFactory)({
  Component: CarouselNavigation,
  mappedArrayProp: 'items'
});
/**
 * A Carousel navigation helps switching between Carousel items.
 */

var _default = (0, _types.withSafeTypeForAs)(CarouselNavigation);

exports.default = _default;
