// @fluentui/react-northstar/dist/commonjs/components/Carousel/CarouselNavigationItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var _accessibility = require("@fluentui/accessibility");

var _utils = require("../../utils");

var _types = require("../../types");

var _Box = _interopRequireDefault(require("../Box/Box"));

var CarouselNavigationItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(CarouselNavigationItem, _UIComponent);

  function CarouselNavigationItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, CarouselNavigationItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(CarouselNavigationItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function (e) {
      (0, _invoke2.default)(_this.props, 'onBlur', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleFocus", function (e) {
      (0, _invoke2.default)(_this.props, 'onFocus', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      performClick: function performClick(event) {
        return !event.defaultPrevented && _this.handleClick(event);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(CarouselNavigationItem, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          variables = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content,
          indicator = _this$props.indicator;
      return (0, _utils.childrenExist)(children) ? children : React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root,
        onBlur: this.handleBlur,
        onFocus: this.handleFocus,
        onClick: this.handleClick
      }, accessibility.attributes.root, _utils.rtlTextContainer.getAttributes({
        forElements: [children]
      }), (0, _utils.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps), unhandledProps), _Box.default.create(indicator, {
        defaultProps: function defaultProps() {
          return {
            className: CarouselNavigationItem.slotClassNames.indicator,
            styles: styles.indicator
          };
        }
      }), _Box.default.create(content, {
        defaultProps: function defaultProps() {
          return {
            as: 'span',
            styles: styles.content
          };
        }
      }));
    }
  }]);
  return CarouselNavigationItem;
}(_utils.UIComponent);

(0, _defineProperty2.default)(CarouselNavigationItem, "displayName", 'CarouselNavigationItem');
(0, _defineProperty2.default)(CarouselNavigationItem, "className", 'ui-carousel__navigationitem');
(0, _defineProperty2.default)(CarouselNavigationItem, "slotClassNames", {
  indicator: "".concat(CarouselNavigationItem.className, "__indicator")
});
(0, _defineProperty2.default)(CarouselNavigationItem, "create", void 0);
(0, _defineProperty2.default)(CarouselNavigationItem, "propTypes", Object.assign({}, _utils.commonPropTypes.createCommon(), {
  active: PropTypes.bool,
  indicator: customPropTypes.shorthandAllowingChildren,
  iconOnly: PropTypes.bool,
  index: PropTypes.number,
  onClick: PropTypes.func,
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  vertical: PropTypes.bool
}));
(0, _defineProperty2.default)(CarouselNavigationItem, "defaultProps", {
  accessibility: _accessibility.tabBehavior,
  as: 'li',
  indicator: {}
});
CarouselNavigationItem.create = (0, _utils.createShorthandFactory)({
  Component: CarouselNavigationItem,
  mappedArrayProp: 'content'
});
/**
 * A CarouselItem is an actionable item within a Carousel.
 */

var _default = (0, _types.withSafeTypeForAs)(CarouselNavigationItem);

exports.default = _default;
