// @fluentui/react-northstar/dist/commonjs/components/SplitButton/SplitButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@fluentui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _positioner = require("../../utils/positioner");

var _types = require("../../types");

var _utils = require("../../utils");

var _SplitButtonToggle = _interopRequireDefault(require("./SplitButtonToggle"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _MenuButton = _interopRequireDefault(require("../MenuButton/MenuButton"));

var _reactComponentRef = require("@fluentui/react-component-ref");

var SplitButton =
/*#__PURE__*/
function (_AutoControlledCompon) {
  (0, _inherits2.default)(SplitButton, _AutoControlledCompon);

  function SplitButton() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, SplitButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(SplitButton)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "targetRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMenuButtonOverrides", function (predefinedProps) {
      return {
        onMenuItemClick: function onMenuItemClick(e, menuItemProps) {
          _this.setState({
            open: false
          });

          (0, _invoke2.default)(_this.props, 'onOpenChange', e, Object.assign({}, _this.props, {
            open: false
          }));
          (0, _invoke2.default)(predefinedProps, 'onMenuItemClick', e, menuItemProps);
          (0, _invoke2.default)(_this.props, 'onMenuItemClick', e, menuItemProps);
        },
        onOpenChange: function onOpenChange(e, popupProps) {
          e.stopPropagation();

          _this.setState({
            open: popupProps.open
          });

          (0, _invoke2.default)(_this.props, 'onOpenChange', e, Object.assign({}, _this.props, {
            open: popupProps.open
          }));
        }
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMenuButtonTriggerOverrides", function (predefinedProps) {
      return {
        onClick: function onClick(e, buttonProps) {
          (0, _invoke2.default)(predefinedProps, 'onClick', e, buttonProps);
          (0, _invoke2.default)(_this.props, 'onMainButtonClick', e, buttonProps);
        },
        onFocus: function onFocus(e, buttonProps) {
          (0, _invoke2.default)(predefinedProps, 'onFocus', e, buttonProps);

          _this.setState({
            isFromKeyboard: (0, _utils.isFromKeyboard)()
          });
        }
      };
    });
    return _this;
  }

  (0, _createClass2.default)(SplitButton, [{
    key: "getInitialAutoControlledState",
    value: function getInitialAutoControlledState() {
      return {
        isFromKeyboard: false,
        open: false
      };
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          button = _this$props.button,
          disabled = _this$props.disabled,
          menu = _this$props.menu,
          primary = _this$props.primary,
          secondary = _this$props.secondary,
          toggleButton = _this$props.toggleButton,
          size = _this$props.size,
          position = _this$props.position,
          align = _this$props.align,
          flipBoundary = _this$props.flipBoundary,
          overflowBoundary = _this$props.overflowBoundary,
          positionFixed = _this$props.positionFixed,
          offset = _this$props.offset,
          unstable_pinned = _this$props.unstable_pinned;

      var trigger = _Button.default.create(button, {
        defaultProps: function defaultProps() {
          return {
            styles: styles.menuButton,
            primary: primary,
            secondary: secondary,
            disabled: disabled
          };
        },
        overrideProps: this.handleMenuButtonTriggerOverrides
      });

      return React.createElement(_reactComponentRef.Ref, {
        innerRef: this.targetRef
      }, React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps), _MenuButton.default.create({}, {
        defaultProps: function defaultProps() {
          return {
            accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.menuButton : undefined,
            menu: menu,
            // Opening is handled manually.
            on: [],
            open: _this2.state.open,
            trigger: trigger,
            target: _this2.targetRef.current,
            position: position,
            align: align,
            flipBoundary: flipBoundary,
            overflowBoundary: overflowBoundary,
            positionFixed: positionFixed,
            offset: offset,
            unstable_pinned: unstable_pinned
          };
        },
        overrideProps: this.handleMenuButtonOverrides
      }), _SplitButtonToggle.default.create(toggleButton, {
        defaultProps: function defaultProps() {
          return Object.assign({
            disabled: disabled,
            primary: primary,
            secondary: secondary,
            size: size
          }, accessibility.attributes.toggleButton);
        },
        overrideProps: function overrideProps(predefinedProps) {
          return {
            onClick: function onClick(e, buttonProps) {
              (0, _invoke2.default)(predefinedProps, 'onClick', e, buttonProps);

              _this2.setState(function (state) {
                var open = !state.open;
                (0, _invoke2.default)(_this2.props, 'onOpenChange', e, Object.assign({}, _this2.props, {
                  open: open
                }));
                return {
                  open: open
                };
              });
            }
          };
        }
      })));
    }
  }]);
  return SplitButton;
}(_utils.AutoControlledComponent);
/**
 * A SplitButton enables users to take one of several related actions, one being dominant and rest being displayed in a menu.
 */


(0, _defineProperty2.default)(SplitButton, "create", void 0);
(0, _defineProperty2.default)(SplitButton, "displayName", 'SplitButton');
(0, _defineProperty2.default)(SplitButton, "Toggle", _SplitButtonToggle.default);
(0, _defineProperty2.default)(SplitButton, "className", 'ui-splitbutton');
(0, _defineProperty2.default)(SplitButton, "propTypes", Object.assign({}, _utils.commonPropTypes.createCommon({
  content: false
}), {
  button: customPropTypes.itemShorthand,
  defaultOpen: PropTypes.bool,
  menu: PropTypes.oneOfType([customPropTypes.itemShorthand, PropTypes.arrayOf(customPropTypes.itemShorthandWithoutJSX)]),
  onMainButtonClick: PropTypes.func,
  onMenuItemClick: PropTypes.func,
  onOpenChange: PropTypes.func,
  open: PropTypes.bool,
  size: customPropTypes.size,
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  toggleButton: customPropTypes.itemShorthand,
  position: PropTypes.oneOf(_positioner.POSITIONS),
  align: PropTypes.oneOf(_positioner.ALIGNMENTS),
  flipBoundary: PropTypes.oneOfType([PropTypes.object, PropTypes.oneOf(['scrollParent', 'window', 'viewport'])]),
  overflowBoundary: PropTypes.oneOfType([PropTypes.object, PropTypes.oneOf(['scrollParent', 'window', 'viewport'])]),
  positionFixed: PropTypes.bool,
  offset: PropTypes.string,
  unstable_pinned: PropTypes.bool
}));
(0, _defineProperty2.default)(SplitButton, "defaultProps", {
  accessibility: _accessibility.splitButtonBehavior,
  toggleButton: {},
  position: 'below',
  align: 'start'
});
(0, _defineProperty2.default)(SplitButton, "autoControlledProps", ['open']);

var _default = (0, _types.withSafeTypeForAs)(SplitButton);

exports.default = _default;
