// @fluentui/react-northstar/dist/commonjs/components/Table/TableRow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var _reactComponentRef = require("@fluentui/react-component-ref");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _TableCell = _interopRequireDefault(require("./TableCell"));

var _utils = require("../../utils");

var _types = require("../../types");

var _accessibility = require("@fluentui/accessibility");

var _styles = require("@fluentui/styles");

var handleVariablesOverrides = function handleVariablesOverrides(variables) {
  return function (predefinedProps) {
    return {
      variables: (0, _styles.mergeComponentVariables)(variables, predefinedProps.variables)
    };
  };
};

var TableRow =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(TableRow, _UIComponent);

  function TableRow() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, TableRow);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(TableRow)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "rowRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      // https://github.com/microsoft/fluent-ui-react/issues/2150
      unsetRowTabbable: function unsetRowTabbable(e) {
        _this.rowRef.current.setAttribute('tabindex', '-1');
      },
      performClick: function performClick(e) {
        _this.handleClick(e);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      if (e.currentTarget === e.target) {
        (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
        e.preventDefault();
      }
    });
    return _this;
  }

  (0, _createClass2.default)(TableRow, [{
    key: "renderCells",
    value: function renderCells(accessibility, variables) {
      var items = this.props.items;
      var cellAccessibility = accessibility.childBehaviors ? accessibility.childBehaviors.cell : undefined;
      return (0, _map2.default)(items, function (item, index) {
        var cellProps = {
          accessibility: cellAccessibility
        };
        var overrideProps = handleVariablesOverrides(variables);
        return _TableCell.default.create(item, {
          defaultProps: function defaultProps() {
            return cellProps;
          },
          overrideProps: overrideProps
        });
      });
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          variables = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var children = this.props.children;
      var hasChildren = (0, _utils.childrenExist)(children);
      return React.createElement(_reactComponentRef.Ref, {
        innerRef: this.rowRef
      }, React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root,
        onClick: this.handleClick
      }, accessibility.attributes.root, unhandledProps, (0, _utils.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), hasChildren && children, !hasChildren && this.renderCells(accessibility, variables)));
    }
  }]);
  return TableRow;
}(_utils.UIComponent);

(0, _defineProperty2.default)(TableRow, "displayName", 'TableRow');
(0, _defineProperty2.default)(TableRow, "className", 'ui-table__row');
(0, _defineProperty2.default)(TableRow, "create", void 0);
(0, _defineProperty2.default)(TableRow, "propTypes", Object.assign({}, _utils.commonPropTypes.createCommon({
  content: false
}), {
  content: customPropTypes.every([customPropTypes.disallow(['children']), PropTypes.oneOfType([PropTypes.arrayOf(customPropTypes.nodeContent), customPropTypes.nodeContent])]),
  items: customPropTypes.collectionShorthand,
  header: PropTypes.bool,
  compact: PropTypes.bool,
  selected: PropTypes.bool
}));
(0, _defineProperty2.default)(TableRow, "defaultProps", {
  accessibility: _accessibility.tableRowBehavior
});
TableRow.create = (0, _utils.createShorthandFactory)({
  Component: TableRow,
  mappedArrayProp: 'items'
});
/**
 * Component represents a single row in a tabular structure
 */

var _default = (0, _types.withSafeTypeForAs)(TableRow);

exports.default = _default;
