// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Attachment/attachmentVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../../../utils");

var _default = function _default(siteVariables) {
  return {
    padding: "".concat((0, _utils.pxToRem)(7), " ").concat((0, _utils.pxToRem)(3), " ").concat((0, _utils.pxToRem)(7), " ").concat((0, _utils.pxToRem)(11)),
    // padding set to 1px less to account for 1px border
    iconSpace: (0, _utils.pxToRem)(12),
    iconSize: (0, _utils.pxToRem)(32),
    borderColor: siteVariables.colors.grey[200],
    borderRadius: (0, _utils.pxToRem)(3),
    backgroundColor: siteVariables.colors.grey[100],
    backgroundColorHover: siteVariables.colors.grey[150],
    textColor: siteVariables.colors.grey[750],
    textColorHover: siteVariables.colors.grey[750],
    boxShadow: siteVariables.shadowLevel1,
    focusBackgroundColor: undefined,
    focusColor: undefined,
    siblingsFocusColor: undefined,
    siblingsHoverColor: undefined,
    progressColor: siteVariables.colors.green[200],
    progressHeight: 4,
    headerFontSize: siteVariables.fontSizes.medium,
    headerFontWeight: siteVariables.fontWeightSemibold,
    headerLineHeight: siteVariables.lineHeightMedium,
    descriptionFontSize: siteVariables.fontSizes.small,
    descriptionFontWeight: siteVariables.fontWeightRegular,
    descriptionLineHeight: siteVariables.lineHeightDefault
  };
};

exports.default = _default;
