// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Attachment/attachmentVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../../../utils");
exports.default = (function (siteVariables) { return ({
    padding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(3) + " " + utils_1.pxToRem(7) + " " + utils_1.pxToRem(11),
    iconSpace: utils_1.pxToRem(12),
    iconSize: utils_1.pxToRem(32),
    borderColor: siteVariables.colors.grey[200],
    borderRadius: utils_1.pxToRem(3),
    backgroundColor: siteVariables.colors.grey[100],
    backgroundColorHover: siteVariables.colors.grey[150],
    textColor: siteVariables.colors.grey[750],
    textColorHover: siteVariables.colors.grey[750],
    boxShadow: siteVariables.shadowLevel1,
    focusBackgroundColor: undefined,
    focusColor: undefined,
    siblingsFocusColor: undefined,
    siblingsHoverColor: undefined,
    progressColor: siteVariables.colors.green[200],
    progressHeight: 4,
    headerFontSize: siteVariables.fontSizes.medium,
    headerFontWeight: siteVariables.fontWeightSemibold,
    headerLineHeight: siteVariables.lineHeightMedium,
    descriptionFontSize: siteVariables.fontSizes.small,
    descriptionFontWeight: siteVariables.fontWeightRegular,
    descriptionLineHeight: siteVariables.lineHeightDefault,
}); });
