// @fluentui/react-northstar/dist/es/themes/teams/components/Attachment/attachmentStyles.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { pxToRem } from '../../../../utils';
import SvgIcon from '../../../../components/SvgIcon/SvgIcon';
import getBorderFocusStyles from '../../getBorderFocusStyles';
import getIconFillOrOutlineStyles from '../../getIconFillOrOutlineStyles';
import Button from '../../../../components/Button/Button';
var attachmentStyles = {
  root: function root(_ref) {
    var _Object$assign, _hover;

    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    var borderFocusStyles = getBorderFocusStyles({
      variables: siteVariables,
      borderRadius: v.borderRadius
    });
    return Object.assign({
      position: 'relative',
      display: 'inline-flex',
      alignItems: 'center',
      width: '100%',
      maxWidth: pxToRem(440),
      minHeight: pxToRem(48),
      padding: v.padding,
      marginBottom: pxToRem(2),
      marginRight: pxToRem(2),
      background: v.backgroundColor,
      color: v.textColor,
      boxShadow: v.boxShadow,
      border: "".concat(siteVariables.borderWidth, " solid ").concat(v.borderColor),
      borderRadius: v.borderRadius
    }, borderFocusStyles, {}, p.actionable && {
      cursor: 'pointer',
      ':focus-visible': Object.assign({}, borderFocusStyles[':focus-visible'], (_Object$assign = {
        backgroundColor: v.focusBackgroundColor,
        color: v.focusColor
      }, _defineProperty(_Object$assign, "& .".concat(Button.className), {
        color: v.siblingsFocusColor
      }), _defineProperty(_Object$assign, "& .".concat(SvgIcon.className), {
        color: v.siblingsFocusColor
      }), _Object$assign)),
      ':hover': (_hover = {
        background: v.backgroundColorHover,
        color: v.textColorHover
      }, _defineProperty(_hover, "& .".concat(Button.className), {
        color: v.siblingsHoverColor
      }), _defineProperty(_hover, "& .".concat(SvgIcon.className), {
        color: v.siblingsHoverColor
      }), _hover)
    });
  },
  content: function content() {
    return {
      flex: 1
    };
  },
  header: function header(_ref2) {
    var v = _ref2.variables;
    return {
      fontSize: v.headerFontSize,
      fontWeight: v.headerFontWeight,
      lineHeight: v.headerLineHeight
    };
  },
  description: function description(_ref3) {
    var v = _ref3.variables;
    return {
      display: 'block',
      fontSize: v.descriptionFontSize,
      fontWeight: v.descriptionFontWeight,
      lineHeight: v.descriptionLineHeight
    };
  },
  icon: function icon(_ref4) {
    var _ref5;

    var v = _ref4.variables;
    return _ref5 = {
      flex: '0 0 auto',
      height: v.iconSize,
      width: v.iconSize
    }, _defineProperty(_ref5, "& .".concat(SvgIcon.className), {
      height: '100%',
      width: '100%',
      '& svg': {
        height: '100%',
        width: '100%'
      }
    }), _defineProperty(_ref5, "marginRight", v.iconSpace), _ref5;
  },
  action: function action(_ref6) {
    var p = _ref6.props,
        v = _ref6.variables,
        siteVariables = _ref6.theme.siteVariables;
    var iconFilledStyles = getIconFillOrOutlineStyles({
      outline: false
    });
    var borderFocusStyles = getBorderFocusStyles({
      variables: siteVariables,
      borderRadius: v.borderRadius
    });
    return Object.assign(_defineProperty({}, "& .".concat(SvgIcon.className), {
      color: v.textColor // this breaks the color change on hover

    }), getIconFillOrOutlineStyles({
      outline: true
    }), {
      ':hover': Object.assign({}, iconFilledStyles, {
        background: 'transparent'
      }),
      ':focus': borderFocusStyles[':focus'],
      ':focus-visible': Object.assign({}, iconFilledStyles, {}, borderFocusStyles[':focus-visible'])
    });
  },
  progress: function progress(_ref7) {
    var p = _ref7.props,
        v = _ref7.variables;
    return {
      transition: 'width 0.2s',
      position: 'absolute',
      display: 'block',
      bottom: 0,
      left: 0,
      maxWidth: '100%',
      height: pxToRem(v.progressHeight),
      background: v.progressColor
    };
  }
};
export default attachmentStyles;
