// @fluentui/react-northstar/dist/dts/src/components/Toolbar/ToolbarDivider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_bindings_1 = require("@fluentui/react-bindings");
var styles_1 = require("@fluentui/styles");
// @ts-ignore
var react_fela_1 = require("react-fela");
var React = require("react");
var types_1 = require("../../types");
var utils_1 = require("../../utils");
var toolbarVariablesContext_1 = require("./toolbarVariablesContext");
var ToolbarDivider = function (props) {
    var context = React.useContext(react_fela_1.ThemeContext);
    var _a = react_bindings_1.useTelemetry(ToolbarDivider.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, className = props.className, design = props.design, styles = props.styles, variables = props.variables;
    var parentVariables = React.useContext(toolbarVariablesContext_1.ToolbarVariablesContext);
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: ToolbarDivider.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(ToolbarDivider.displayName, {
        className: ToolbarDivider.className,
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: styles_1.mergeComponentVariables(parentVariables, variables),
        }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(ToolbarDivider.handledProps, props);
    var element = React.createElement(ElementType, tslib_1.__assign({}, getA11yProps('root', tslib_1.__assign(tslib_1.__assign({}, unhandledProps), { className: classes.root }))));
    setEnd();
    return element;
};
ToolbarDivider.className = 'ui-toolbar__divider';
ToolbarDivider.displayName = 'ToolbarDivider';
ToolbarDivider.propTypes = utils_1.commonPropTypes.createCommon();
ToolbarDivider.handledProps = Object.keys(ToolbarDivider.propTypes);
ToolbarDivider.create = utils_1.createShorthandFactory({ Component: ToolbarDivider, mappedProp: 'content' });
/**
 * A ToolbarDivider is a non-actionable element that visually segments Toolbar items.
 */
exports.default = types_1.withSafeTypeForAs(ToolbarDivider);
